<a href="https://colab.research.google.com/github/graporris/trabajo-final-analisis-de-datos-alkemy/blob/main/TrabajoFinal.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#PROYECTO FINAL

##Importación e instalación de librerías

In [9]:
!pip install pandas==2.2.2
!pip install numpy
!pip install lxml
import pandas as pd
import numpy as np
import os

##Importación de datasets

In [10]:
df_customers = pd.read_csv("/content/drive/MyDrive/Analisis de datos - Alkemy/Datasets originales/ecommerce_customers_dataset.csv")
df_customers.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

In [11]:
df_order_items = pd.read_csv("/content/drive/MyDrive/Analisis de datos - Alkemy/Datasets originales/ecommerce_order_items_dataset.csv")
df_order_items.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [12]:
df_order_payments = pd.read_csv("/content/drive/MyDrive/Analisis de datos - Alkemy/Datasets originales/ecommerce_order_payments_dataset.csv")
df_order_payments.head()

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

In [13]:
df_orders = pd.read_csv("/content/drive/MyDrive/Analisis de datos - Alkemy/Datasets originales/ecommerce_orders_dataset.csv")
df_orders.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

In [14]:
df_products = pd.read_csv("/content/drive/MyDrive/Analisis de datos - Alkemy/Datasets originales/ecommerce_products_dataset.csv")
df_products.head()

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

##Análisis de la estructura de los datasets

###Función analisisDatosDF

In [15]:
#Definición de la función para hacer el análisis

def analisisDatosDF(df) :
  print("Tamaño DataFrame: \n", df.shape)
  print("---------------------\n")
  print("Análisis del tipo de datos:")
  print(df.info())
  print("---------------------\n")
  print("Recuento de nulos: \n", df.isnull().sum())
  print("---------------------\n")
  print("Recuento de datos duplicados: \n", df.duplicated().sum())

###Customers

In [16]:
analisisDatosDF(df_customers)

Tamaño DataFrame: 
 (99441, 5)
---------------------

Análisis del tipo de datos:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB
None
---------------------

Recuento de nulos: 
 customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64
---------------------

Recuento de datos duplicados: 
 0


###Products

In [17]:
analisisDatosDF(df_products)

Tamaño DataFrame: 
 (32951, 9)
---------------------

Análisis del tipo de datos:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB
None
---------------------

Recuento de nulos: 
 product_id                      0
product_category_name         610
product_name_l

###Order_items

In [18]:
analisisDatosDF(df_order_items)

Tamaño DataFrame: 
 (112650, 7)
---------------------

Análisis del tipo de datos:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB
None
---------------------

Recuento de nulos: 
 order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64
---------------------

Recuento de datos duplicad

###Order_payments

In [19]:
analisisDatosDF(df_order_payments)

Tamaño DataFrame: 
 (103886, 5)
---------------------

Análisis del tipo de datos:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB
None
---------------------

Recuento de nulos: 
 order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64
---------------------

Recuento de datos duplicados: 
 0


###Orders

In [20]:
analisisDatosDF(df_orders)

Tamaño DataFrame: 
 (99441, 8)
---------------------

Análisis del tipo de datos:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB
None
---------------------

Recuento de nulos: 
 order_id                            0
customer_id                         0
order_status                        0
order_purchase_ti

##Transformación dataframes

Se convierten los datos al tipo que sea más conveniente y se homogenizan.
En aquellos dataframes que hay nulos si la proporcion no es muy grande es factible eliminar dichas filas.

###Función transformarFechas

Convierte las columnas incluidas en columnas_fechas a datetime

In [21]:
def transformarFechas(df, columnas_fechas) :

  df[columnas_fechas] = df[columnas_fechas].apply(pd.to_datetime)

  return df

###Función transformarString

Convierte las columnas incluidas en columnas_string a string

In [22]:
def transformarString(df, columnas_string) :

  for columna in columnas_string :
    df[columna] = df[columna].astype("string")

  return df

###Función convertirMayusculas

Convierte las columnas incluidas en columna a mayúsculas

In [23]:
def convertirMayusculas(df, columna) :

  df[columna] = df[columna].str.upper()

  return df

###Función convertirMinusculas

Convierte las columnas incluidas en columna a minúsculas

In [24]:
def convertirMinusculas(df, columna) :

  df[columna] = df[columna].str.lower()

  return df

###Customers

In [25]:
#Consersión a string de las columnas incluidas en columnas_string_customers
columnas_string_customers = ["customer_id", "customer_unique_id", "customer_zip_code_prefix", "customer_city", "customer_state"]

df_customers = transformarString(df_customers, columnas_string_customers)

df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  string
 1   customer_unique_id        99441 non-null  string
 2   customer_zip_code_prefix  99441 non-null  string
 3   customer_city             99441 non-null  string
 4   customer_state            99441 non-null  string
dtypes: string(5)
memory usage: 3.8 MB


In [26]:
#Conversión a minusculas de la columna customer_city
df_customers = convertirMinusculas(df_customers, "customer_city")

#Conversión a mayusculas de la columna customer_state
df_customers = convertirMayusculas(df_customers, "customer_state")


##Products

In [27]:
#Conversión a string de las columnas incluidas en columnas_products
columnas_string_products = ["product_id", "product_category_name"]

df_products = transformarString(df_products, columnas_string_products)

df_products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  string 
 1   product_category_name       32341 non-null  string 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), string(2)
memory usage: 2.3 MB


In [28]:
#Conversión a minusculas de la columna product_category_name
df_products = convertirMinusculas(df_products, "product_category_name")

In [29]:
#Análisis del porcentaje de nulos en cada columna
porcentaje_nulos_productos = (df_products.isnull().sum()/df_products.shape[0])*100

porcentaje_nulos_productos

product_id                    0.000000
product_category_name         1.851234
product_name_lenght           1.851234
product_description_lenght    1.851234
product_photos_qty            1.851234
product_weight_g              0.006070
product_length_cm             0.006070
product_height_cm             0.006070
product_width_cm              0.006070
dtype: float64

In [30]:
#Filtra columnas que contienen un porcentaje de nulos menor al 10%
columnas_a_mantener_productos = porcentaje_nulos_productos[porcentaje_nulos_productos < 10].index

#Eliminación de las filas con nulos menor al 10%
df_products.dropna(subset = columnas_a_mantener_productos, inplace = True)

#Verificación
df_products.isnull().sum()

product_id                    0
product_category_name         0
product_name_lenght           0
product_description_lenght    0
product_photos_qty            0
product_weight_g              0
product_length_cm             0
product_height_cm             0
product_width_cm              0
dtype: int64

##Order_items

In [31]:
#Conversión a string de las columnas incluidas en columnas_string_order_items
columnas_string_order_items = ["order_id", "product_id", "seller_id"]

df_order_items = transformarString(df_order_items, columnas_string_order_items)

#Conversión a fecha de la columna shipping_limit_date
columnas_fechas_order_items = ['shipping_limit_date']

df_order_items = transformarFechas(df_order_items, columnas_fechas_order_items)

df_order_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   order_id             112650 non-null  string        
 1   order_item_id        112650 non-null  int64         
 2   product_id           112650 non-null  string        
 3   seller_id            112650 non-null  string        
 4   shipping_limit_date  112650 non-null  datetime64[ns]
 5   price                112650 non-null  float64       
 6   freight_value        112650 non-null  float64       
dtypes: datetime64[ns](1), float64(2), int64(1), string(3)
memory usage: 6.0 MB


In [32]:
#Análisis de nulos
porcentaje_nulos_order_items = (df_order_items.isnull().sum()/df_order_items.shape[0])*100

porcentaje_nulos_order_items

order_id               0.0
order_item_id          0.0
product_id             0.0
seller_id              0.0
shipping_limit_date    0.0
price                  0.0
freight_value          0.0
dtype: float64

##Order_payments

In [33]:
# #   Column                Non-Null Count   Dtype
# ---  ------                --------------   -----
#  0   order_id              103886 non-null  object
#  1   payment_sequential    103886 non-null  int64
#  2   payment_type          103886 non-null  object
#  3   payment_installments  103886 non-null  int64
#  4   payment_value         103886 non-null  float64

#Conversión a string de las columnas
columnas_string_order_payments = ["order_id", "payment_type"]

df_order_payments = transformarString(df_order_payments, columnas_string_order_payments)

df_order_payments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  string 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  string 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), string(2)
memory usage: 4.0 MB


In [34]:
#Conversión a minusculas de la columna payment_type
df_order_payments = convertirMinusculas(df_order_payments, "payment_type")

In [35]:
#Análisis de nulos
porcentaje_nulos_order_payments = (df_order_payments.isnull().sum()/df_order_payments.shape[0])*100

porcentaje_nulos_order_payments

order_id                0.0
payment_sequential      0.0
payment_type            0.0
payment_installments    0.0
payment_value           0.0
dtype: float64

##Orders

In [36]:
#Conversión a string columnas
columnas_string_orders = ["order_id", "customer_id", "order_status"]

df_orders = transformarString(df_orders, columnas_string_orders)

#Conversión a fecha columnas
columnas_fechas_orders = ["order_purchase_timestamp", "order_approved_at", "order_delivered_carrier_date",
                               "order_delivered_customer_date", "order_estimated_delivery_date"]

df_orders = transformarFechas(df_orders, columnas_fechas_orders)

df_orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       99441 non-null  string        
 1   customer_id                    99441 non-null  string        
 2   order_status                   99441 non-null  string        
 3   order_purchase_timestamp       99441 non-null  datetime64[ns]
 4   order_approved_at              99281 non-null  datetime64[ns]
 5   order_delivered_carrier_date   97658 non-null  datetime64[ns]
 6   order_delivered_customer_date  96476 non-null  datetime64[ns]
 7   order_estimated_delivery_date  99441 non-null  datetime64[ns]
dtypes: datetime64[ns](5), string(3)
memory usage: 6.1 MB


In [37]:
#Conversión a minúsculas de la columna order_status
df_orders = convertirMinusculas(df_orders, "order_status")

In [38]:
#Análisis de nulos
porcentaje_nulos_orders = (df_orders.isnull().sum()/df_orders.shape[0])*100

porcentaje_nulos_orders

order_id                         0.000000
customer_id                      0.000000
order_status                     0.000000
order_purchase_timestamp         0.000000
order_approved_at                0.160899
order_delivered_carrier_date     1.793023
order_delivered_customer_date    2.981668
order_estimated_delivery_date    0.000000
dtype: float64

In [39]:
#Filtra columnas que contienen un porcentaje de nulos menor al 10%
columnas_a_mantener_orders = porcentaje_nulos_orders[porcentaje_nulos_orders < 10].index

#Eliminación de las filas con nulos menor al 10%
df_orders.dropna(subset = columnas_a_mantener_orders, inplace = True)

#Verificación
df_orders.isnull().sum()

order_id                         0
customer_id                      0
order_status                     0
order_purchase_timestamp         0
order_approved_at                0
order_delivered_carrier_date     0
order_delivered_customer_date    0
order_estimated_delivery_date    0
dtype: int64

##Análisis dataframes

Análisis de los datos que contienen los dataframes para extrar la infromación relevante

###Función analisisDataFrames

In [40]:
def analisisDF(df) :

  print("Análisis de variables cuantitativas: \n")
  columnas_numericas = df.select_dtypes(include=["int", "float"]).columns  #Identificación de columnas numéricas

  if columnas_numericas.empty :   #Si no tiene columnas numericas lo indica en un mensaje
    print("No posee")
  else :
    print(df[columnas_numericas].describe())  #Si tiene columnas numéricas hace el análisis

  print("----------------------\n")
  print("Análisis de variables cualitativas: \n")
  columnas_string = df.select_dtypes(include=["string"]).columns #Identificación de columnas string

  if columnas_string.empty :  #Si no tiene columnas string lo indica en un mensaje
    print("No posee")
  else :
    print(df[columnas_string].describe())  #Si tiene columnas string hace el análisis



###Customers

In [41]:
analisisDF(df_customers)

Análisis de variables cuantitativas: 

No posee
----------------------

Análisis de variables cualitativas: 

                             customer_id                customer_unique_id  \
count                              99441                             99441   
unique                             99441                             96096   
top     06b8999e2fba1a1fbc88172c00ba8bc7  8d50f5eadf50201ccdcedfb9e2ac8455   
freq                                   1                                17   

       customer_zip_code_prefix customer_city customer_state  
count                     99441         99441          99441  
unique                    14994          4119             27  
top                       22790     sao paulo             SP  
freq                        142         15540          41746  


In [42]:
df_customers["customer_state"].value_counts()

customer_state
SP    41746
RJ    12852
MG    11635
RS     5466
PR     5045
SC     3637
BA     3380
DF     2140
ES     2033
GO     2020
PE     1652
CE     1336
PA      975
MT      907
MA      747
MS      715
PB      536
PI      495
RN      485
AL      413
SE      350
TO      280
RO      253
AM      148
AC       81
AP       68
RR       46
Name: count, dtype: Int64

Estados con mayor cantidad de clientes:

*   Sao Pablo
*   Rio de Janeiro
*   Minas Gerais
*   Rio Grande do Sul
*   Paraná



In [43]:
df_customers["customer_city"].value_counts()

customer_city
sao paulo            15540
rio de janeiro        6882
belo horizonte        2773
brasilia              2131
curitiba              1521
                     ...  
palmares paulista        1
pires ferreira           1
vicente dutra            1
alagoa nova              1
eugenio de castro        1
Name: count, Length: 4119, dtype: Int64

Ciudades con mayor cantidad de clientes:

*   Sao Pablo
*   Rio de Janeiro
*   Belo Horizonte
*   Brasilia
*   Curitiba


###Porducts

In [44]:
analisisDF(df_products)

Análisis de variables cuantitativas: 

       product_name_lenght  product_description_lenght  product_photos_qty  \
count         32340.000000                32340.000000        32340.000000   
mean             48.476592                  771.492393            2.188961   
std              10.245699                  635.124831            1.736787   
min               5.000000                    4.000000            1.000000   
25%              42.000000                  339.000000            1.000000   
50%              51.000000                  595.000000            1.000000   
75%              57.000000                  972.000000            3.000000   
max              76.000000                 3992.000000           20.000000   

       product_weight_g  product_length_cm  product_height_cm  \
count      32340.000000       32340.000000       32340.000000   
mean        2276.956586          30.854545          16.958813   
std         4279.291845          16.955965          13.636115  

In [45]:
df_products["product_category_name"].value_counts()

product_category_name
cama_mesa_banho                  3029
esporte_lazer                    2867
moveis_decoracao                 2657
beleza_saude                     2444
utilidades_domesticas            2335
                                 ... 
fashion_roupa_infanto_juvenil       5
casa_conforto_2                     5
pc_gamer                            3
seguros_e_servicos                  2
cds_dvds_musicais                   1
Name: count, Length: 73, dtype: Int64

Categoría de productos más vendidos:

*   Cama mesa banho
*   Esporte lazer
*   moveis decoracao
*   Beleza saude
*   Utilidades domesticas


Categorías de productos menos vendidos:
* PC gamer
* Seguros e servicos
* cd y dvd musicales


Las medidas promedio(en cm) de los productos son:         
* Length: 30.854545          
* Height: 16.958813
* Width: 23.208596

El peso promedio de los productos(en gramos) es:
* Weight: 2276.956586


##Order items

In [46]:
analisisDF(df_order_items)

Análisis de variables cuantitativas: 

       order_item_id          price  freight_value
count  112650.000000  112650.000000  112650.000000
mean        1.197834     120.653739      19.990320
std         0.705124     183.633928      15.806405
min         1.000000       0.850000       0.000000
25%         1.000000      39.900000      13.080000
50%         1.000000      74.990000      16.260000
75%         1.000000     134.900000      21.150000
max        21.000000    6735.000000     409.680000
----------------------

Análisis de variables cualitativas: 

                                order_id                        product_id  \
count                             112650                            112650   
unique                             98666                             32951   
top     8272b63d03f5f79c56e9e4120aec44ef  aca2eb7d00ea1a7b8ebd4e68314663af   
freq                                  21                               527   

                               seller_id  
count 

In [47]:
df_order_items["price"].sum()

13591643.700000003

Han tenido un total de 112650 ventas, lo que generó un ingreso de 	R$13591643.70.

El precio promedio de los productos vendidos es de R$120, pero el 50% tienen un precio menor a R$74.99 y el 75% menor a R$134.90.

El costo promedio del envío es de	R$19.99

###Order pyments

In [48]:
analisisDF(df_order_payments)

Análisis de variables cuantitativas: 

       payment_sequential  payment_installments  payment_value
count       103886.000000         103886.000000  103886.000000
mean             1.092679              2.853349     154.100380
std              0.706584              2.687051     217.494064
min              1.000000              0.000000       0.000000
25%              1.000000              1.000000      56.790000
50%              1.000000              1.000000     100.000000
75%              1.000000              4.000000     171.837500
max             29.000000             24.000000   13664.080000
----------------------

Análisis de variables cualitativas: 

                                order_id payment_type
count                             103886       103886
unique                             99440            5
top     fa65dad1b0e818e3ccc5cb0e39231352  credit_card
freq                                  29        76795


In [49]:
df_order_payments["payment_type"].value_counts()

payment_type
credit_card    76795
boleto         19784
voucher         5775
debit_card      1529
not_defined        3
Name: count, dtype: Int64

EL principal medio de pago es la tarjeta de crédito

###Orders

In [50]:
analisisDF(df_orders)

Análisis de variables cuantitativas: 

No posee
----------------------

Análisis de variables cualitativas: 

                                order_id                       customer_id  \
count                              96461                             96461   
unique                             96461                             96461   
top     e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
freq                                   1                                 1   

       order_status  
count         96461  
unique            2  
top       delivered  
freq          96455  


In [51]:
df_orders["order_status"].value_counts()

order_status
delivered    96455
canceled         6
Name: count, dtype: Int64

De 96461 ventas, sólo 6 fueron canceladas.

##Nuevos Dataframes

A partir de los dataset originales y despues de trabajar con las columnas que los componen vamos a generar dos nuevos DataFrames que exportaremos a csv para poder utilizarlos y generar un reporte con visualizaciones

In [52]:
if not os.path.exists('dataset_generados'):   #Crea la carpeta para guardar los datasets si no existe
    os.makedirs('dataset_generados')

###Order payments

In [53]:
df_merge_orders_payments = pd.merge(df_orders, df_order_payments, how = "inner", on= "order_id")

df_merge_orders_payments.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
4  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   

  order_status order_purchase_timestamp   order_approved_at  \
0    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
2    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
3    delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
4    delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   
2          2017-10-04 19:55:00           2017-10-10 21:25:13   
3          2018-07-26 14:31:00           2018-08-07 15:27:45   
4          2018-08-08 13:50:00           2018-08-17 18:06:29   

  order_estimated_delivery_date  payment_sequential payment_type  \
0                    2017-10-18                   1  credit_card   
1                    2017-10-18                   3      voucher   
2                    2017-10-18                   2      voucher   
3                    2018-08-13                   1       boleto   
4                    2018-09-04                   1  credit_card   

   payment_installments  payment_value  
0                     1          18.12  
1                     1           2.00  
2                     1          18.59  
3                     1         141.46  
4                     3         179.12

In [54]:
df_merge_orders_payments.drop(['customer_id','order_approved_at','order_delivered_carrier_date','order_delivered_customer_date','order_estimated_delivery_date'], axis=1, inplace=True) #Columnas que vamos a mantener

df_merge_orders_payments.head()

order_id order_status order_purchase_timestamp  \
0  e481f51cbdc54678b7cc49136f2d6af7    delivered      2017-10-02 10:56:33   
1  e481f51cbdc54678b7cc49136f2d6af7    delivered      2017-10-02 10:56:33   
2  e481f51cbdc54678b7cc49136f2d6af7    delivered      2017-10-02 10:56:33   
3  53cdb2fc8bc7dce0b6741e2150273451    delivered      2018-07-24 20:41:37   
4  47770eb9100c2d0c44946d9cf07ec65d    delivered      2018-08-08 08:38:49   

   payment_sequential payment_type  payment_installments  payment_value  
0                   1  credit_card                     1          18.12  
1                   3      voucher                     1           2.00  
2                   2      voucher                     1          18.59  
3                   1       boleto                     1         141.46  
4                   1  credit_card                     3         179.12

In [55]:
df_merge_orders_payments.to_csv("dataset_generados/orders_payments.csv") #Exporta el dataframe nuevo

###Order items - Products

In [56]:
df_merge_order_items_products = pd.merge(df_order_items, df_products, how = "inner", on= "product_id")
df_merge_order_items_products.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

  shipping_limit_date   price  freight_value product_category_name  \
0 2017-09-19 09:45:35   58.90          13.29            cool_stuff   
1 2017-05-03 11:05:13  239.90          19.93              pet_shop   
2 2018-01-18 14:48:30  199.00          17.87      moveis_decoracao   
3 2018-08-15 10:10:18   12.99          12.79            perfumaria   
4 2017-02-13 13:57:51  199.90          18.14    ferramentas_jardim   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 58.0                       598.0                 4.0   
1                 56.0                       239.0                 2.0   
2                 59.0                       695.0                 2.0   
3                 42.0                       480.0                 1.0   
4                 59.0                       409.0                 1.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             650.0               28.0                9.0              14.0  
1           30000.0               50.0               30.0              40.0  
2            3050.0               33.0               13.0              33.0  
3             200.0               16.0               10.0              15.0  
4            3750.0               35.0               40.0              30.0

In [57]:
df_merge_order_items_products.drop(['product_id','seller_id','shipping_limit_date'], axis=1, inplace=True) #Columnas que vamos a mantener
df_merge_order_items_products.head()

order_id  order_item_id   price  freight_value  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   58.90          13.29   
1  00018f77f2f0320c557190d7a144bdd3              1  239.90          19.93   
2  000229ec398224ef6ca0657da4fc703e              1  199.00          17.87   
3  00024acbcdf0a6daa1e931b038114c75              1   12.99          12.79   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1  199.90          18.14   

  product_category_name  product_name_lenght  product_description_lenght  \
0            cool_stuff                 58.0                       598.0   
1              pet_shop                 56.0                       239.0   
2      moveis_decoracao                 59.0                       695.0   
3            perfumaria                 42.0                       480.0   
4    ferramentas_jardim                 59.0                       409.0   

   product_photos_qty  product_weight_g  product_length_cm  product_height_cm  \
0                 4.0             650.0               28.0                9.0   
1                 2.0           30000.0               50.0               30.0   
2                 2.0            3050.0               33.0               13.0   
3                 1.0             200.0               16.0               10.0   
4                 1.0            3750.0               35.0               40.0   

   product_width_cm  
0              14.0  
1              40.0  
2              33.0  
3              15.0  
4              30.0

In [58]:
df_merge_order_items_products.to_csv("dataset_generados/order_items_products.csv")  #Exporta el dataframe nuevo

###Custormer-orders

In [60]:
df_merge_customers_orders = pd.merge(df_customers, df_orders, how = "inner", on = "customer_id" )

df_merge_customers_orders.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

  customer_zip_code_prefix          customer_city customer_state  \
0                    14409                 franca             SP   
1                     9790  sao bernardo do campo             SP   
2                     1151              sao paulo             SP   
3                     8775        mogi das cruzes             SP   
4                    13056               campinas             SP   

                           order_id order_status order_purchase_timestamp  \
0  00e7ee1b050b8499577073aeb2a297a1    delivered      2017-05-16 15:05:35   
1  29150127e6685892b6eab3eec79f59c7    delivered      2018-01-12 20:48:24   
2  b2059ed67ce144a36e2aa97d2c9e9ad2    delivered      2018-05-19 16:07:45   
3  951670f92359f4fe4a63112aa7306eba    delivered      2018-03-13 16:06:38   
4  6b7d50bd145f6fc7f33cebabd7e49d0f    delivered      2018-07-29 09:51:30   

    order_approved_at order_delivered_carrier_date  \
0 2017-05-16 15:22:12          2017-05-23 10:47:57   
1 2018-01-12 20:58:32          2018-01-15 17:14:59   
2 2018-05-20 16:19:10          2018-06-11 14:31:00   
3 2018-03-13 17:29:19          2018-03-27 23:22:42   
4 2018-07-29 10:10:09          2018-07-30 15:16:00   

  order_delivered_customer_date order_estimated_delivery_date  
0           2017-05-25 10:35:35                    2017-06-05  
1           2018-01-29 12:41:19                    2018-02-06  
2           2018-06-14 17:58:51                    2018-06-13  
3           2018-03-28 16:04:25                    2018-04-10  
4           2018-08-09 20:55:48                    2018-08-15

In [64]:
df_merge_customers_orders.drop(["customer_unique_id","order_status"], axis=1, inplace=True)



KeyError: "['customer_unique_id', 'order_status'] not found in axis"

In [65]:
df_merge_customers_orders.head()

customer_id customer_zip_code_prefix  \
0  06b8999e2fba1a1fbc88172c00ba8bc7                    14409   
1  18955e83d337fd6b2def6b18a428ac77                     9790   
2  4e7b3e00288586ebd08712fdd0374a03                     1151   
3  b2b6027bc5c5109e529d4dc6358b12c3                     8775   
4  4f2d8ab171c80ec8364f7c12e35b23ad                    13056   

           customer_city customer_state                          order_id  \
0                 franca             SP  00e7ee1b050b8499577073aeb2a297a1   
1  sao bernardo do campo             SP  29150127e6685892b6eab3eec79f59c7   
2              sao paulo             SP  b2059ed67ce144a36e2aa97d2c9e9ad2   
3        mogi das cruzes             SP  951670f92359f4fe4a63112aa7306eba   
4               campinas             SP  6b7d50bd145f6fc7f33cebabd7e49d0f   

  order_purchase_timestamp   order_approved_at order_delivered_carrier_date  \
0      2017-05-16 15:05:35 2017-05-16 15:22:12          2017-05-23 10:47:57   
1      2018-01-12 20:48:24 2018-01-12 20:58:32          2018-01-15 17:14:59   
2      2018-05-19 16:07:45 2018-05-20 16:19:10          2018-06-11 14:31:00   
3      2018-03-13 16:06:38 2018-03-13 17:29:19          2018-03-27 23:22:42   
4      2018-07-29 09:51:30 2018-07-29 10:10:09          2018-07-30 15:16:00   

  order_delivered_customer_date order_estimated_delivery_date  
0           2017-05-25 10:35:35                    2017-06-05  
1           2018-01-29 12:41:19                    2018-02-06  
2           2018-06-14 17:58:51                    2018-06-13  
3           2018-03-28 16:04:25                    2018-04-10  
4           2018-08-09 20:55:48                    2018-08-15

In [66]:
df_merge_customers_orders.to_csv("dataset_generados/customer_orders.csv")

Costumer-order items